### PR TITLE
ListView(getMentor) & Mocha setting

### PR DESCRIPTION
**Pull request by SarahhhC**
sprint for this Wednesday! Please comment me!

@nlpenguin @ethkim @KaylaKwon @ifndefdeadmau5 @heartjigi

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,8 +5,6 @@ services:
   - mongodb
 before_script:
   - npm install -g gulp
-  - gulp &
-  - sleep 5
```
Comment (ethkim) [inline .travis.yml]: In another PR (I think kyungah's?), this was set to 10. You can just sync and will get that reflected when she merges hers. :)

File: gulpfile.js [hunk 2]
```diff
@@ -1,15 +1,14 @@
-const gulp = require('gulp');
-const tape = require('gulp-tape');
-const faucet = require('faucet');
+const apidoc = require('gulp-apidoc');
```
Comment (ethkim) [inline gulpfile.js]: Thank you!! Thank you!!

File: src/server/config/mongoose.js [hunk 3]
```diff
@@ -5,6 +5,6 @@ import '../models/users.model';
 
 export default function () {
   mongoose.Promise = global.Promise;
-  const db = mongoose.connect('mongodb://localhost/yoda');
+  const db = mongoose.connect('mongodb://localhost:27017/yoda');
```
Comment (ethkim) [inline src/server/config/mongoose.js]: Why this port #? Is this the default port # for mongodb? If not, we should also update the command to bring up mongodb in the wiki or the README.md file I added to src/

File: src/server/controllers/users.controller.js [hunk 4]
```diff
@@ -34,3 +12,15 @@ export function getAll(req, res, next) {
     }
   });
 }
+
+// Get all user list except logined user
+export function getMentorList(req, res, next) {
+  user.find({ 'email' : { $ne : 'session@yoda.com' }},  // if session_auth add, 'session' => req.session.email
```
Comment (ethkim) [inline src/server/controllers/users.controller.js]: Replace the comment to:

//TODO: Once session_auth is implemented, replace the email address with req.session.email.

Comment (ethkim) [inline src/server/controllers/users.controller.js]: Actually, can we replace email with user_id? The user ID is automatically generated (i think?) each time a user registers, so we don't need to rely on email address.

File: src/server/controllers/users.controller.js [hunk 5]
```diff
@@ -34,3 +12,15 @@ export function getAll(req, res, next) {
...
+    { email: 0, userId: 0, age: 0, region: 0, skills: 0, regDate: 0 }, (err, doc) => {
```
Comment (ethkim) [inline src/server/controllers/users.controller.js]: This is super important. This subset of fields from user object is the set that gets exposed to Node (and thus all subsequent flows). Write that in the comment above this line.

File: src/server/models/users.model.js [hunk 6]
```diff
@@ -2,13 +2,12 @@
...
 import mongoose from 'mongoose';
 
-/**
+/*
  * Define the schema model for user.
  */
 
 const Schema = mongoose.Schema;
 let userSchema = new Schema({
-  userId: String,
```
Comment (ethkim) [inline src/server/models/users.model.js]: Why was this deleted?? Even though email address is enforced to be unique, user ID (numeric) should be there.

File: src/server/controllers/users.controller.js [hunk 7]
```diff
@@ -2,28 +2,6 @@
...
-  email: 'abcds@fsdss.com',
-  name: '홍길동',
-  age: 25,
-  field: 'SW Engineer',
-  region: 'Seoul',
-  skills: 'JAVA',
-};
-
-// Add a mock data into response.
-export function insertMockData(req, res, next) {
```
Comment (ethkim) [inline src/server/controllers/users.controller.js]: why remove this? because we now have the sign up flow complete?

File: src/server/controllers/users.controller.js (hunk 4 above)
Comment (SarahhhC) [inline src/server/controllers/users.controller.js]: I understood that email address is userId, the same thing because we login with Facebook or Linked!

File: src/test/fixtures/loginedUser.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// Sample data for test getMentorList (session not added)
+export default {
+  loginedUserData: {
```
Comment (ethkim) [inline src/test/fixtures/loginedUser.js]: ->loggedInUserData

And also change the name of the file.

File: src/test/controllers/users.controller.test.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+import sampleUser from '../fixtures/loginedUser';
```
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: loggedInUser

File: src/test/controllers/users.controller.test.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+/*
+ * Test for users.controller.
+ */
+
+let assert = require('assert');
+let should = require('should');
+let user = require('mongoose').model('user');
+
+// Test for getMentorList method.
```
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: I think we can remove this line. It's in the wrong place (it should be between line 15 and 16), but line 16 already states the method name.

File: src/test/controllers/users.controller.test.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    it('should create a new User with fake session', function (done) {
+      var u = sampleUser.loginedUserData;
+      user.create(u, function (err, createdUser) {
+        should.not.exist(err);
+        createdUser.name.should.equal('session');
+        done();
+      });
+    });
+    
+    // Need to access getMentorList method
```
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: ->TODO: Add a test to access getMentorList method.

File: src/test/controllers/users.controller.test.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    // Need to check not existing session user
```
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: ->TODO: Add a test to check for non-existent user session.

File: src/test/controllers/users.controller.test.js (hunk 11 above)
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: // TODO: Add a test to access getMentorList method.

File: src/test/controllers/users.controller.test.js (hunk 12 above)
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: // TODO: Add a test to check for non-existent user session.

File: src/test/controllers/users.controller.test.js [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    // Need to check existing all users except session user
```
Comment (ethkim) [inline src/test/controllers/users.controller.test.js]: // TODO: Add a test to check if all users are returned except the user in the session.

File: src/test/fixtures/loginedUser.js [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+    email: 'session@yoda.com',
+    name: 'session',
+  },
+};
```
Comment (ethkim) [inline src/test/fixtures/loginedUser.js]: Add an empty line below.

File: src/server/controllers/users.controller.js (hunk 5 above)
Comment (SarahhhC) [inline src/server/controllers/users.controller.js]: So you mean I need to add some comments about which fields would show? Or need to pass all fields?

File: src/test/models/users.model.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+let user = require('mongoose').model('user');
+
+describe('Users: models', function () {
+  describe('#create()', function () {
+    it('should create a new User', function (done) {
+      // Create a User object to pass to User.create()
+      var u = {
+        name: 'yoda'
+      };
+      user.create(u, function (err, createdUser) {
```
Comment (ethkim) [inline src/test/models/users.model.js]: This is still creating new users (and deleting right after) to the production database. The link I sent out earlier today should contain how to set up without having to touch the prod database. Could we add a TODO to do something like that?

An easy way to do that without much work would be 1) create a new mongoDb (instead of yoda, maybe yoda-test) as part of the test flow, 2) reset the database (drop all tables), and 3) run the tests. Then you're testing against a test db all the time. Of course, it's an overkill, but just a thought :)

File: src/test/models/users.model.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+        should.not.exist(err);
+        createdUser.name.should.equal('yoda');
```
Comment (ethkim) [inline src/test/models/users.model.js]: instead of directly checking for createdUser, get out of this user.create(), and do a user.find(), and check if "yoda" is returned.

File: src/test/models/users.model.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+    });
+  
+    it('should delete new User', function (done) {
+      // Create a User object to pass to User.create()
+      var u = {
+        name: 'yoda'
+      };
+      user.find(u).remove(function() {
+        done();
+      });
```
Comment (ethkim) [inline src/test/models/users.model.js]: You're not checking if the user is removed. Do a find() after remove() and check if nothing is returned.

File: src/test/utils.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+/*
+ * Prepare for mongodb connection to test.
+ */
+
+let mongoose = require('mongoose');
+
+beforeEach(function (done) {
+  function clearDB() {
+    /*
+     * This part is hided because it always cleanDB even when dev-mode,
```
Comment (ethkim) [inline src/test/utils.js]: I think we can bring this back if we use test-specific db (e.g. yoda-test) for all the tests.

File: src/server/models/users.model.js (hunk 6 above)
Comment (KaylaKwon) [inline src/server/models/users.model.js]: @ethkim If there is no other option, the object that saved in collection gets `_id` from MongoDB. It's unique and can be found faster than custom numeric id (like AUTO_INCREMENT index in MySQL).

https://docs.mongodb.com/manual/reference/method/ObjectId/

File: src/server/controllers/users.controller.js (hunk 4 above)
Comment (ethkim) [inline src/server/controllers/users.controller.js]: The userID I mean here is a numeric value, assigned to each user. That would be Yoda specific user ID which we issue, and not the FB/LI ids.

File: src/server/controllers/users.controller.js (hunk 5 above)
Comment (ethkim) [inline src/server/controllers/users.controller.js]: On line between 18 and 19, write a comment that says:

// TODO: Longer term, we should migrate to a UserSummary object 
// that contains subset of fields. For now, we return only the following 
// fields from user object:
//
// - email
// - userId
// - age
// - region
// - skills
// - regDate

This is probably the least you can do. You're writing the api, so this is just to help whoever that wants to use the user objects in other part of the system. 

I think in the longer term, we'd have to define UserSummary object that has only a subset of fields from User object, and get passed around the system. If someone wants more detailed fields for the user, they can do another db lookup. 

It's hairy :) that's why I initially suggested using the entire user object until the performance starts becoming an issue. But what you have here is ok for now.

File: src/test/models/users.model.js (hunk 15 above)
Comment (SarahhhC) [inline src/test/models/users.model.js]: Creating new database for test would be better for later or now. I'll add TODO in trello!

**Comment by ethkim** [comment]
Thank you!
Please make adjustments for the comments above. I think the biggest change I'm suggesting is to use user_id as part of the user model.

File: src/server/models/users.model.js (hunk 6 above)
Comment (ethkim) [inline src/server/models/users.model.js]: Yes, we can use that: ensure we copy _id into our custom field user_id though. Otherwise it's a weird hashCode for the object which will end up being difficult to work with in some other context.

File: src/server/config/mongoose.js (hunk 3 above)
Comment (SarahhhC) [inline src/server/config/mongoose.js]: Yes mongodb's default port # is 27017...!

File: src/server/controllers/users.controller.js (hunk 7 above)
Comment (SarahhhC) [inline src/server/controllers/users.controller.js]: This insertMockData method was for initial back-end test, so it's not needed for now. Maybe sign up flow will replace this!

File: src/server/controllers/users.controller.js (hunk 5 above)
Comment (SarahhhC) [inline src/server/controllers/users.controller.js]: Ok actually 'email: 0' means 'I'll not return email field',
so I'll change this as { userId: 1, email: 0, name: 1, age: 0, field: 1, region: 0, skills: 0, regDate: 0 }...

File: src/test/controllers/users.controller.test.js [hunk 19]
```diff
@@ -11,24 +11,17 @@ let assert = require('assert');
...
-    // Need to access getMentorList method
-    // Need to check not existing session user
-    // Need to check existing all users except session user
-  
-    it('should delete new User with fake session', function (done) {
-      user.find({ name: 'session' }).remove(function() {
+        // TODO: Add a test to access getMentorList method.
+        // TODO: Add a test to check for non-existent user session.
+        // TODO: Add a test to check if all users are returned except the user in the session.
+        createdUser.remove();
```
Comment (SarahhhC) [inline src/test/controllers/users.controller.test.js]: @ethkim 
I thought this is clear than find created user and delete...
Is this looks better?

File: src/server/controllers/users.controller.js [hunk 20]
```diff
@@ -13,10 +13,18 @@ export function getAll(req, res, next) {
   });
 }
 
-// Get all user list except logined user
+// Get all user list except loggedIn user
```
Comment (ethkim) [inline src/server/controllers/users.controller.js]: "logged in"

File: src/server/models/users.model.js [hunk 21]
```diff
@@ -8,6 +8,7 @@ import mongoose from 'mongoose';
 
 const Schema = mongoose.Schema;
 let userSchema = new Schema({
+  userId: String,
```
Comment (ethkim) [inline src/server/models/users.model.js]: Is this string because the object ID is string?

File: src/test/models/users.model.js [hunk 22]
```diff
@@ -6,29 +6,20 @@ import '../../server/models/users.model';
...
-      });
-    });
-  
-    it('should delete new User', function (done) {
-      // Create a User object to pass to User.create()
-      var u = {
-        name: 'yoda'
-      };
-      user.find(u).remove(function() {
+        createdUser.remove();
```
Comment (ethkim) [inline src/test/models/users.model.js]: In all of these tests, create() should actually do a lookup from the db (rather than looking at the createdUser) so the object is actually inserted in the db. Once that verification is done, then we can do createUser.remove(). 

Same goes for line 24 of users.controller.test.js; we should just create a few users A and B, then pull up all users using getMentorList() method, and check if the returned list contains precisely A and B (and not the current session user), and finally remove A and B. 

Can you add these in the TODO comments? Without these, we're merely testing for user.create() method which isn't even our own code.

**Comment by ethkim** [comment]
Left few more comments, but you can put most of them as TODO and get back to them later. Please fix as suggested, and then you can merge. LGTM.

File: src/server/models/users.model.js (hunk 21 above)
Comment (SarahhhC) [inline src/server/models/users.model.js]: Oh... so we don't need userId if we need Object id. It will be duplicated. The _id is created automatically and added as _id field...
Did you mean saving userId as _id value???

Comment (SarahhhC) [inline src/server/models/users.model.js]: _id is created when user instance created so we can use _id value as id. But if we add userId field and copy the _id value after user instance created, _id and userId field would be duplicated!

File: src/server/models/users.model.js [hunk 23]
```diff
@@ -8,6 +8,7 @@ import mongoose from 'mongoose';
...
   email: String,
   name: String,
   age: Number,
```
Comment (heartjigi) [inline src/server/models/users.model.js]: Are this elements  mock up ?
we need data of company, job position, etc...

File: src/test/models/users.model.js (hunk 22 above)
Comment (SarahhhC) [inline src/test/models/users.model.js]: In method create(), createdUser is exact same instance(collection) that created right after, so I thought find() is not necessary(already verified maybe) and using createdUser is more clear...

But if testDB created, we don't need to consider about removing because it'll always clear DB, right?

Comment (ethkim) [inline src/test/models/users.model.js]: I see. Then at the beginning of every test case, we need a routine to clear the entire DB. One of the URLs I sent about Mocha was doing that by setting up an environment function that gets run before every unit test. For now, I thought we should just do that manually.

File: src/server/models/users.model.js (hunk 23 above)
Comment (ethkim) [inline src/server/models/users.model.js]: @SarahhhC Let's add those fields in the schema even if we don't have that data automatically. So at least the client folks can start rendering it. 

@heartjigi : in case those data fields are empty, we need a graceful way to display the empty fields in the UI. Make sure the client doesn't crash or anything for those cases :)

File: src/server/models/users.model.js (hunk 21 above)
Comment (ethkim) [inline src/server/models/users.model.js]: I'm not sure if there's any flow in the product to get "gildong" as the user ID. We don't have a sign up flow, so all the info is coming from fb / li. So the "gildong" is merely coming from the mock data.

_id itself is present for every single object, right? So I think it's ok to duplicate them as user_id for user objects. I'm just not sure if that _id is persistent : if the mongodb crashes and we'll have to recover from a backup, would we still maintain that object _id? Things like that. 

Let's just use _id as the user_id for now, and in the mean time, let's find out how other systems are setting up user_ids in their applications.

File: src/server/models/users.model.js (hunk 23 above)
Comment (SarahhhC) [inline src/server/models/users.model.js]: ok, I'll add it!

File: src/server/models/users.model.js (hunk 21 above)
Comment (SarahhhC) [inline src/server/models/users.model.js]: Oh, I didn't consider mongodb crash or persistent...
This time I'll use _id instead of userId and I'll find out other systems about user_id!

**Comment by KaylaKwon** [comment]
LGTM

File: src/server/config/mongoose.js (hunk 3 above)
Comment (nlpenguin) [inline src/server/config/mongoose.js]: Use a constant for the port number.
```
const mongodb_port = 27017
```

**Comment by nlpenguin** [comment]
LGTM

**Comment by ethkim** [comment]
There are conflicts u need resolve. LGTM Thanks Sarah!

File: src/test/controllers/users.controller.test.js [hunk 24]
```diff
@@ -0,0 +1,29 @@
+'use strict';
```
Comment (fguy) [inline src/test/controllers/users.controller.test.js]: ES6 doesn't require this: http://stackoverflow.com/questions/31685262/not-recommended-to-write-out-use-strict-with-es6

File: src/test/controllers/users.controller.test.js [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+
+import sampleUser from '../fixtures/loggedInUserData';
+import '../../server/models/users.model';
+
+/*
+ * Test for users.controller.
+ */
+
+let assert = require('assert');
```
Comment (fguy) [inline src/test/controllers/users.controller.test.js]: This should be replaced to 

import assert from 'assert';

require is deprecated in es6.

File: src/test/controllers/users.controller.test.js [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+let should = require('should');
```
Comment (fguy) [inline src/test/controllers/users.controller.test.js]: This line should be replaced to module import.

File: src/test/controllers/users.controller.test.js [hunk 27]
```diff
@@ -0,0 +1,29 @@
...
+let user = require('mongoose').model('user');
```
Comment (fguy) [inline src/test/controllers/users.controller.test.js]: ditto

File: src/test/controllers/users.controller.test.js [hunk 28]
```diff
@@ -0,0 +1,29 @@
...
+
+describe('Test for users.controller', function() {
+  describe('#getMentorList()', function() {
+    it('should create a new User with fake session', function (done) {
+      var u = sampleUser.loggedInUserData;
```
Comment (fguy) [inline src/test/controllers/users.controller.test.js]: Do not use var: http://programmers.stackexchange.com/questions/274342/is-there-any-reason-to-use-the-var-keyword-in-es6

File: src/test/models/users.model.js [hunk 29]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+import '../../server/models/users.model';
+
+/*
+ * Test for mongodb create.
+ */
+
+const should = require('should');
+let user = require('mongoose').model('user');
```
Comment (fguy) [inline src/test/models/users.model.js]: This should be
const user ...
since user hasn't been reassigned.

File: src/test/models/users.model.js [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+let u = {
```
Comment (fguy) [inline src/test/models/users.model.js]: ditto

File: src/test/utils.js [hunk 31]
```diff
@@ -0,0 +1,36 @@
+'use strict';
```
Comment (fguy) [inline src/test/utils.js]: This line is not needed.

File: src/test/utils.js [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+
+/*
+ * Prepare for mongodb connection to test.
+ */
+
+let mongoose = require('mongoose');
```
Comment (fguy) [inline src/test/utils.js]: Replace this line to module import.

File: src/test/utils.js [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+
+beforeEach(function (done) {
+  function clearDB() {
```
Comment (fguy) [inline src/test/utils.js]: Use arrow function for function declaration.

File: src/test/utils.js [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  
+  if (mongoose.connection.readyState === 0) {
+    mongoose.connect('mongodb://localhost:27017/yoda', function (err) {
+      if (err) {
+        throw err;
+      }
+      return clearDB();
+    });
+  } else {
```
Comment (fguy) [inline src/test/utils.js]: You don't need else here.

File: src/test/utils.js [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+    return clearDB();
+  }
+});
+
+afterEach(function (done) {
```
Comment (fguy) [inline src/test/utils.js]: function (done) { ... } can be
(done) => { ... }

File: src/test/controllers/users.controller.test.js (hunk 24 above)
Comment (SarahhhC) [inline src/test/controllers/users.controller.test.js]: Really? then I'll modify every file...!

**Comment by ifndefdeadmau5** [comment]
LGTM
